torchvision/models/detection/fcos.py

```diff
@@ -0,0 +1,701 @@
+        >>> # load a pre-trained model for classification and return
+        >>> # only the features
+        >>> backbone = torchvision.models.mobilenet_v2(pretrained=True).features
+        >>> # FCOS needs to know the number of
+        >>> # output channels in a backbone. For mobilenet_v2, it's 1280
+        >>> # so we need to add it here
+        >>> backbone.out_channels = 1280
+        >>>
+        >>> # let's make the network generate 5 x 3 anchors per spatial
+        >>> # location, with 5 different sizes and 3 different aspect
+        >>> # ratios. We have a Tuple[Tuple[int]] because each feature
+        >>> # map could potentially have different sizes and
+        >>> # aspect ratios
+        >>> anchor_generator = AnchorGenerator(
+        >>>     sizes=((8,), (16,), (32,), (64,), (128,)),
+        >>>     aspect_ratios=((1.0,),)
+        >>> )
+        >>>
+        >>> # put the pieces together inside a FCOS model
+        >>> model = FCOS(
+        >>>     backbone,
+        >>>     num_classes=80,
+        >>>     anchor_generator=anchor_generator,
+        >>> )
+        >>> model.eval()
+        >>> x = [torch.rand(3, 300, 400), torch.rand(3, 500, 400)]
+        >>> predictions = model(x)
+    """
+
+    __annotations__ = {
+        "box_coder": det_utils.BoxLinearCoder,
+    }
+
+    def __init__(
+        self,
+        backbone,
+        num_classes,
+        # transform parameters
+        min_size=800,
+        max_size=1333,
+        image_mean=None,
+        image_std=None,
+        # Anchor parameters
+        anchor_generator=None,
+        head=None,
+        center_sampling_radius=1.5,
+        score_thresh=0.2,
+        nms_thresh=0.5,
+        detections_per_img=300,
+        topk_candidates=1000,
+    ):
+        super().__init__()
+        _log_api_usage_once(self)
+
+        if not hasattr(backbone, "out_channels"):
+            raise ValueError(
+                "backbone should contain an attribute out_channels "
+                "specifying the number of output channels (assumed to be the "
+                "same for all the levels)"
+            )
+        self.backbone = backbone
+
+        assert isinstance(anchor_generator, (AnchorGenerator, type(None)))
+
+        if anchor_generator is None:
+            anchor_sizes = ((8,), (16,), (32,), (64,), (128,))  # equal to strides of multi-level feature map
+            aspect_ratios = ((1.0,),) * len(anchor_sizes)  # set only one anchor
+            anchor_generator = AnchorGenerator(anchor_sizes, aspect_ratios)
+        self.anchor_generator = anchor_generator
+        assert self.anchor_generator.num_anchors_per_location()[0] == 1
+
+        if head is None:
+            head = FCOSHead(backbone.out_channels, anchor_generator.num_anchors_per_location()[0], num_classes)
+        self.head = head
+
+        self.box_coder = det_utils.BoxLinearCoder(normalize_by_size=True)
+
+        if image_mean is None:
+            image_mean = [0.485, 0.456, 0.406]
```

Comment:
The `[0.485, 0.456, 0.406]` are typical values used in TorchVision see here: https://github.com/pytorch/vision/blame/main/docs/source/models.rst#L142

Also accordingly to the documentation, these values are estimated on a subset of ImageNet and which subset was used is lost. See #1439 and #1965.

We should be OK using these values instead of the ones on the original repo to ensure consistency across TorchVision.

Reply:
@jdsgomes for detection2 or maskrcnn-benchmark，the image is not normalized to [0, 1] but use [0, 255], so the mean and std are different，another difference is that the std is included into the first conv weights of resnet for detection2.

Reply:
> @datumbox The implementation is much similar to detectron2. We use the tricks as descibed as the original paper:
> 
> * add the centerness brach on the regression branch instead of classification branch;
> * center sampling strategy (radis = 1.5);
> * use GIoU loss for the regression loss;
> * normalize the regression targets by the stride of feature map, but do not use `Scale` layer on the regression branch;
> 
> there is also three minor difference compared to the offical implementation ([AdelaiDet](https://github.com/aim-uofa/AdelaiDet)):
> 
> * normalize the regression loss  by the number of positive sample (similar to other losses) instead of sum of ctrness_targets;
> * use the square root of the multiplication of classification score and centerness to threshold, instead of classification score. so the score_thresh is much greater than normal setting (0.2 vs 0.05).
> * the nms thresh is 0.5 as normal instead of 0.6 used in paper.
> 
> But I think they has negligible influence on the detection result.
> 
> When you are free, you can check them.

@jdsgomes

Reply:
@jdsgomes OK.